File: sdk/ruby/lib/telecom_sdk.rb
require "httparty"
require "json"
require "websocket-client-simple"
require "uri"

require_relative "telecom_sdk/auth"
require_relative "telecom_sdk/http_client"
require_relative "telecom_sdk/subscribers_api"
require_relative "telecom_sdk/usage_api"
require_relative "telecom_sdk/payments_api"
require_relative "telecom_sdk/rating_plans_api"
require_relative "telecom_sdk/system_api"
require_relative "telecom_sdk/graphql_api"

module TelecomSDK
  class Error < StandardError; end
  class AuthenticationError < Error; end
  class APIError < Error; end
  class NetworkError < Error; end
  class ValidationError < Error; end
  class RateLimitError < Error; end
  class ServerError < Error; end

  class Client
    attr_reader :config, :subscribers, :usage, :payments, :rating_plans, :system, :graphql

    def initialize(config = {})
      @config = Config.new(config)
      @auth_provider = AuthProvider.new(api_key: @config.api_key, jwt_secret: @config.jwt_secret)
      @http_client = HTTPClient.new(
        @config.api_url,
        @auth_provider,
        timeout: @config.timeout,
        max_retries: @config.max_retries,
        retry_delay: @config.retry_delay,
        enable_logging: @config.enable_logging
      )

      # Initialize API modules
      @subscribers = SubscriberAPI.new(@http_client)
      @usage = UsageAPI.new(@http_client)
      @payments = PaymentAPI.new(@http_client)
      @rating_plans = RatingPlanAPI.new(@http_client)
      @system = SystemAPI.new(@http_client)
      @graphql = GraphQLAPI.new(@http_client)
    end

    # Authentication methods

    def generate_jwt_token(user_id, expiry_hours = 24, additional_claims = {})
      @auth_provider.generate_jwt_token(user_id, expiry_hours, additional_claims)
    end

    def validate_jwt_token(token)
      @auth_provider.validate_jwt_token(token)
    end

    # Cleanup

    def close
      @http_client.close
    end
  end

  class Config
    attr_accessor :api_url, :api_key, :jwt_secret, :timeout, :max_retries, :retry_delay, :enable_logging

    def initialize(config = {})
      @api_url = config[:api_url] || "http://localhost:8000"
      @api_key = config[:api_key]
      @jwt_secret = config[:jwt_secret]
      @timeout = config[:timeout] || 30
      @max_retries = config[:max_retries] || 3
      @retry_delay = config[:retry_delay] || 1
      @enable_logging = config[:enable_logging] || false
    end
  end
end

File: sdk/typescript/src/telecom-sdk.ts
import { AuthProvider } from './auth';
import { 
  HTTPClient, 
  SubscriberAPI, 
  UsageAPI, 
  PaymentAPI, 
  RatingPlanAPI, 
  SystemAPI, 
  GraphQLAPI 
} from './api';
import { WebSocketClient } from './websocket';
import { Alert, RealTimeUsageUpdate, WebSocketMessage } from './types';

export class TelecomSDK {
  private static instance: TelecomSDK;
  private authProvider: AuthProvider;
  private apiClient: HTTPClient;
  private ws: WebSocketClient;
  
  // API modules
  public subscribers: SubscriberAPI;
  public usage: UsageAPI;
  public payments: PaymentAPI;
  public ratingPlans: RatingPlanAPI;
  public system: SystemAPI;
  public graphql: GraphQLAPI;

  private constructor(config: {
    baseURL: string;
    apiKey?: string;
    jwtSecret?: string;
    timeout?: number;
    maxRetries?: number;
    retryDelay?: number;
    enableLogging?: boolean;
    websocketURL?: string;
  }) {
    const authConfig: any = {};
    if (config.apiKey !== undefined) authConfig.apiKey = config.apiKey;
    if (config.jwtSecret !== undefined) authConfig.jwtSecret = config.jwtSecret;
    
    this.authProvider = new AuthProvider(authConfig);
    
    this.apiClient = new HTTPClient(
      config.baseURL,
      this.authProvider,
      config.timeout || 30000,
      config.maxRetries || 3,
      config.retryDelay || 1000,
      config.enableLogging || false
    );
    
    this.ws = new WebSocketClient(
      config.websocketURL || config.baseURL.replace('http', 'ws'),
      config.enableLogging || false
    );
    
    // Initialize API modules
    this.subscribers = new SubscriberAPI(this.apiClient);
    this.usage = new UsageAPI(this.apiClient);
    this.payments = new PaymentAPI(this.apiClient);
    this.ratingPlans = new RatingPlanAPI(this.apiClient);
    this.system = new SystemAPI(this.apiClient);
    this.graphql = new GraphQLAPI(this.apiClient);
  }

  static initialize(config: {
    baseURL: string;
    apiKey?: string;
    jwtSecret?: string;
    timeout?: number;
    maxRetries?: number;
    retryDelay?: number;
    enableLogging?: boolean;
    websocketURL?: string;
  }): TelecomSDK {
    if (!TelecomSDK.instance) {
      TelecomSDK.instance = new TelecomSDK(config);
    }
    return TelecomSDK.instance;
  }

  static getInstance(): TelecomSDK {
    if (!TelecomSDK.instance) {
      throw new Error('TelecomSDK not initialized. Call TelecomSDK.initialize() first.');
    }
    return TelecomSDK.instance;
  }

  // Authentication methods
  generateJWTToken(userId: string, expiryHours: number = 24, additionalClaims: Record<string, any> = {}): string {
    return this.authProvider.generateJWTToken(userId, expiryHours, additionalClaims);
  }

  validateJWTToken(token: string): any {
    return this.authProvider.validateJWTToken(token);
  }

  // WebSocket methods
  async connectWebSocket(): Promise<void> {
    await this.ws.connect();
  }

  disconnectWebSocket(): void {
    this.ws.disconnect();
  }

  get isWebSocketConnected(): boolean {
    return this.ws.isConnected;
  }

  // Convenience methods
  onUsageUpdate(callback: (update: RealTimeUsageUpdate) => void): void {
    this.ws.on('usage_update', (message: WebSocketMessage) => {
      callback(message.data as RealTimeUsageUpdate);
    });
  }

  onAlert(callback: (alert: Alert) => void): void {
    this.ws.on('alert', (message: WebSocketMessage) => {
      callback(message.data as Alert);
    });
  }

  // Cleanup
  destroy(): void {
    this.disconnectWebSocket();
    this.apiClient.close();
  }
}

// Export types for external use
export * from './types';
export { AuthProvider, AuthConfig, JWTClaims } from './auth';
export * from './api';
export { WebSocketClient } from './websocket';
